File: LeetCode/392. Is Subsequence.js
// https://leetcode.com/problems/is-subsequence/

const { test } = require('../test');

const isSubsequence = (substr, str) => {
  const strLength = str.length;
  const substrLength = substr.length;

  if (strLength < substrLength) return false;

  let index = 0;

  for (let i = 0; i < strLength; i++) {
    if (str[i] === substr[index]) {
      index++;
    }

    if (index === substrLength) {
      return true;
    }
  }

  return index === substrLength;
};

test(isSubsequence, [
  {
    input: ['abc', 'ahbgdc'],
    expected: true,
  },
  {
    input: ['axc', 'ahbgdc'],
    expected: false,
  },
]);

File: LeetCode/392. Is Subsequence.py
# https://leetcode.com/problems/is-subsequence/

from test import test


class Solution:
    def isSubsequence(
        self,
        substring: str,
        string: str,
    ) -> bool:
        stringLength = len(string)
        substringLength = len(substring)

        if substringLength == 0:
            return True

        if substringLength > stringLength:
            return False

        index = 0

        for char in string:
            if char == substring[index]:
                index += 1

            if index == substringLength:
                return True

        return index == substringLength


solution = Solution()


test(
    solution.isSubsequence,
    [
        {
            "input": ["abc", "ahbgdc"],
            "expected": True,
        },
        {"input": ["axc", "ahbgdc"], "expected": False},
        {
            "input": [
                "",
                "ahbgdc",
            ],
            "expected": True,
        },
    ],
)
